build/gen_stub.php

```diff
@@ -629,26 +662,332 @@ private function getFlagsAsString(): string
 
         return $flags;
     }
+}
+
+class PropertyInfo
+{
+    /** @var PropertyName */
+    public $name;
+    /** @var int */
+    public $flags;
+    /** @var bool */
+    public $isKnownName;
+    /** @var bool */
+    public $isDeprecated;
+    /** @var Type|null */
+    public $type;
+    /** @var Expr|null */
+    public $defaultValue;
+
+    public function __construct(PropertyName $name, int $flags, bool $isKnownName, bool $isDeprecated, ?Type $type, ?Expr $defaultValue)
+    {
+        $this->name = $name;
+        $this->flags = $flags;
+        $this->isKnownName = $isKnownName;
+        $this->isDeprecated = $isDeprecated;
+        $this->type = $type;
+        $this->defaultValue = $defaultValue;
+    }
 
     public function discardInfoForOldPhpVersions(): void {
-        $this->return->type = null;
-        foreach ($this->args as $arg) {
-            $arg->type = null;
-            $arg->defaultValue = null;
+        $this->type = null;
+    }
+
+    public function getDeclaration(): string {
+        $code = "\\\n";
+
+        $propertyName = $this->name->property;
+
+        $defaultValueConstant = false;
+        if ($this->defaultValue === null) {
+            $defaultValue = null;
+            $defaultValueType = "undefined";
+        } else {
+            $evaluator = new ConstExprEvaluator(
+                function (Expr $expr) use (&$defaultValueConstant) {
+                    if ($expr instanceof Expr\ConstFetch) {
+                        $defaultValueConstant = true;
+                        return null;
+                    }
+
+                    throw new Exception("Property $this->name has an unsupported default value");
+                }
+            );
+            $defaultValue = $evaluator->evaluateDirectly($this->defaultValue);
+            $defaultValueType = gettype($defaultValue);
+        }
+
+        if ($defaultValueConstant) {
+            echo "Skipping code generation for property $this->name, because it has a constant default value\n";
```

Comment:
I felt this is the cleanest way to handle constant values. I could generate code for these (especially if the PHP constant is the same as the C constant), but it would still feel nasty.